### PR TITLE
Update advanced composition code to avoid a numeric overflow.

### PR DESCRIPTION
**Pull request by matthewclegg**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)

### CODE REVIEW

File: src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/Composition.kt [hunk 1]
```diff
@@ -62,8 +70,8 @@ object Composition {
       var deltaI = 0.0f
       for (l in 0..i - 1) {
         deltaI +=
-          coeff(k, l).toFloat() *
-            (exp(epsilon * (k - l).toFloat()) - exp(epsilon * (k - 2 * i + l).toFloat()))
+          exp(logBinomial(k, l) + (epsilon * (k - l).toFloat())) -
+            exp(logBinomial(k, l) + epsilon * (k - 2 * i + l).toFloat())
       }
       deltaI /= (1.0f + exp(epsilon)).pow(k)
```
Comment (jiayu-google) [inline src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/Composition.kt:76]: Wondering if there's a benefit to also take log of denominator and put it inside exp?

**Review by jiayu-google — APPROVED** [review]
Thanks Matt!  LGTM but left a minor comment.

**Review by uakyol — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)* status: all files reviewed, 1 unresolved discussion (waiting on @matthewclegg)

**Review by matthewclegg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)* status: all files reviewed, 1 unresolved discussion (waiting on @jiayu-google)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/Composition.kt` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576#-N325TP47ptk7J1du7Vj-r1-76:-N3QHKlBBiKZXTQU6A6Z:b-1nsl7f) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9acce27baaf37d0d65ce425d70d17c0479d09e77/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/Composition.kt#L76)):*
<details><summary><i>Previously, jiayu-google wrote…</i></summary><blockquote>

Wondering if there's a benefit to also take log of denominator and put it inside exp?
</blockquote></details>

Thanks for the review!  For the logBinomial, there is a clear benefit because the code produced an integer overflow if k was bigger than about 75.  Since deltaI is a float, it does not suffer from this problem, and I don't see a clear benefit to making this change.  Could you explain why you think this would be helpful?

**Review by jiayu-google — APPROVED** [review]

**Review by duliomatos1 — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)* status: all files reviewed, 1 unresolved discussion (waiting on @jiayu-google)

**Review by matthewclegg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @jiayu-google, and @uakyol)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/Composition.kt` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576#-N325TP47ptk7J1du7Vj-r1-76:-N3b7rOy1RIlvq_G5OJp:b5vohrh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/9acce27baaf37d0d65ce425d70d17c0479d09e77/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/Composition.kt#L76)):*
<details><summary><i>Previously, matthewclegg (Matthew Clegg) wrote…</i></summary><blockquote>

Thanks for the review!  For the logBinomial, there is a clear benefit because the code produced an integer overflow if k was bigger than about 75.  Since deltaI is a float, it does not suffer from this problem, and I don't see a clear benefit to making this change.  Could you explain why you think this would be helpful?
</blockquote></details>

Turns out you were right about this.  I should have listened!

**Comment by jiayu-google** [comment]
Thanks Matt! 
Somehow, some tests were not successful. Could you check them?

**Review by matthewclegg — COMMENTED** [review]
It looks like the file PrivacyCharge.kt has mysteriously disappeared from the repository.  I am going to check with Ugur about it.

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576#-:-N3eIABPAkg5rPAKidDZ:bj2v0qi)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @jiayu-google, and @uakyol)

**Review by jiayu-google — APPROVED** [review]
LGTM

**Review by uakyol — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @jiayu-google)

**Review by duliomatos1 — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576#-:-N43ETdH4XtLUkApSHi-:bg7igkv)* status: all files reviewed, 1 unresolved discussion (waiting on @jiayu-google)

**Review by matthewclegg — COMMENTED** [review]
> Dismissed @jiayu-google from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/576)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @matthewclegg)

**Pull request merged**